chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -37,109 +37,245 @@ MallocStatus PinnedMemoryAllocator::Malloc(void** ptr, size_t bytesize) {
+        for (auto free_list_it = free_list.begin(); free_list_it != free_list.end();) {
+            std::unique_ptr<Chunk>& chunk = *free_list_it;
+            if (chunk->next() == nullptr && chunk->prev() == nullptr) {
+                allocator_->Free(chunk->ptr());
+                free_list_it = free_list.erase(free_list_it);
+            } else {
+                ++free_list_it;
+            }
         }
+        // Compact free_bins
+        if (free_list.empty()) {
+            free_bins_it = free_bins_.erase(free_bins_it);
+        } else {
+            ++free_bins_it;
+        }
+    }
+    if (free_bins_.empty()) {
+        free_bins_.clear();
     }
-    free_bins_.clear();
 }
 
 void* MemoryPool::Malloc(size_t bytesize) {
     if (bytesize == 0) {
         return nullptr;
     }
 
-    size_t index = (bytesize - 1) / kAllocationUnitSize;
+    // TODO(niboshi): Currently the deleter of allocated memory assumes that
+    // the memory is stored in the memory pool (in `in_use_`), but this may not hold if some exception is thrown before it is stored.
+    // `std::lock_guard` and `in_use_.emplace` are the sources of possible exceptions.
+
+    size_t allocation_size = GetAllocationSize(bytesize);
+    std::unique_ptr<Chunk> chunk{nullptr};
 
-    void* ptr = nullptr;
     {
         std::lock_guard<std::mutex> lock{free_bins_mutex_};
-
-        if (index < free_bins_.size()) {
-            std::vector<void*>& free_list = free_bins_[index];
-            if (!free_list.empty()) {
-                ptr = free_list.back();
-                CHAINERX_ASSERT(ptr != nullptr);
-                free_list.pop_back();
-            }
-        }
+        chunk = PopFromFreeList(allocation_size);
     }
 
-    // TODO(niboshi): Currently the deleter of allocated memory assumes that
-    // the memory is stored in the memory pool (in `in_use_`), but this may not hold if some exception is thrown before it is stored.
-    // `std::lock_guard` and `in_use_.emplace` are the sources of possible exceptions.
-    if (ptr == nullptr) {
-        size_t allocation_size = (index + 1) * kAllocationUnitSize;
+    if (chunk != nullptr) {
+        std::unique_ptr<Chunk> remaining = chunk->Split(allocation_size);
+        if (remaining != nullptr) {
+            std::lock_guard<std::mutex> lock{free_bins_mutex_};
+            PushIntoFreeList(std::move(remaining));
+        }
+    } else {
+        void* ptr{nullptr};
         CudaSetDeviceScope scope{device_index_};
         MallocStatus status = allocator_->Malloc(&ptr, allocation_size);
         if (status == MallocStatus::kErrorMemoryAllocation) {
             FreeUnusedBlocks();
             status = allocator_->Malloc(&ptr, allocation_size);
             if (status == MallocStatus::kErrorMemoryAllocation) {
                 // TODO(sonots): Include total pooled bytes in the error message
-                throw OutOfMemoryError{bytesize};
+                throw OutOfMemoryError{allocation_size};
```

Comment:
Why this change? I think `bytesize` is what this class tries to allocate. `allocation_size` does not mean anything to the caller.

Reply:
`allocation_size` is just an internal implementation. It does not represent any meaningful number to users. It does not necessarily represent a low-level allocation size either (which is passed to cudaMemAlloc for example), because allocators are abstracted.

Presenting this number could confuse users. (Actually I was confused when I saw this number).

Reply:
Fixed.